catppuccin-papirus-folders: 0-unstable-2023-08-02 -> 0-unstable-2024-08-06

Sewer56 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Updates `catppuccin-papirus-folders` from [fcf9673](https://github.com/catppuccin/papirus-folders/commit/fcf96737fffc343a1bf129732c37d19f2d77fa5c) to [f83671d](https://github.com/catppuccin/papirus-folders/commit/f83671d17ea67e335b34f8028a7e6d78bca735d7) (latest).

The most important change is the removal of the `papirus-folders` script from upstream.

Upstream now says `fetch it from papirus-folders`, so I'm doing an URL fetch on a specific commit (latest at time of writing) to ensure reproducability.

## Need Help

In this PR I had to set `dontFixup = true`, which is likely undesirable.

Help is requested here. To the best of my awareness [per nixpkgs docs](https://nixos.org/manual/nixpkgs/stable/#ssec-fixup-phase), there are no binaries, man pages, etc. in build output, yet the fixup step locks up. Removing fixup doesn't seem like good practice. So help is requested.

More details here:

- https://github.com/NixOS/nixpkgs/issues/426952

I'm not sure how to debug this, as I've only got 3 days experience with Nix(OS) ***(I haven't even set up my machine fully yet!!)***, so help would be appreciated.

Aside from that, works as expected, if fixup is skipped.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

MattSturgeon commented on pkgs/by-name/ca/catppuccin-papirus-folders/package.nix:
```diff
@@ -31,27 +33,43 @@ let
+      # Copy the upstream papirus-folders script (pinned), per papirus-folders documentation
+      cp ${papirus-folders-script} ./papirus-folders
+      chmod +x ./papirus-folders
       patchShebangs ./papirus-folders
```
Likewise, this can be done within `fetchurl`'s `postFetch` script.

Although, I'm guessing a copy or symlink is still needed if this script is used later in the build. A symlink would be preferred, if possible.

Sewer56 replied:
It was needed, unfortunately I couldn't call `patchShebangs` from postFetch.
I was able to clean up a bit by fetchUrl with executable, thanks.

MattSturgeon replied:
It should be possible if we also add whatever hook provides `patchShebangs` to fetchurl's `nativeBuildInputs`. Not strictly needed, though.

MattSturgeon replied:
Ah, I see:

```
error: fixed-output derivations must not reference store paths: '/nix/store/998spbf7ifzzc4iwi6i4i0132dcyh8hd-papirus-folders.drv' references 1 distinct paths, e.g. '/nix/store/gkwbw9nzbkbz298njbn3577zmrnglbbi-bash-5.3p0'
```

A fixed-output derivation cannot depend on a _non_-fixed-output derivation, in this case bash.

Makes sense.

MattSturgeon commented on pkgs/by-name/ca/catppuccin-papirus-folders/package.nix:
```diff
@@ -31,27 +33,43 @@ let
     };
 
-    nativeBuildInputs = [ gtk3 ];
+    # This gets stuck on fixup for some reason. Please help.
+    dontFixup = true;
```
I can try and investigate. What errors were you seeing?

Sewer56 replied:
A lockup, described in:
- https://github.com/NixOS/nixpkgs/issues/426952

It would run a `find` command during the fixup phase, which would not find anything; and thus the script would error out, but without showing an error in the end.

Unfortunately, verbose mode, nor `--show-trace` gave me any useful info here.

MattSturgeon replied:
When I build I get `fixupPhase completed in 21 minutes 16 seconds`, so the fixup phase does run, but it is _very_ slow.

I suspect the sheer number of files and symlinks is the issue... `231 directories, 122993 files`

MattSturgeon replied:
I agree there shouldn't be anything too important in the fixup phase. I don't remember what the default fixup phase includes, but the most important stuff relates to executables, which we don't have in this package.

Disabling the fixup phase should be fine if we tweak the comment to mention it a) struggles with the sheer number of files in the package and b) probably doesn't have anything too important for this package.

If it turns out there is stuff we need in the fixup phase, (maybe some permission/owner fixes?), then we can instead explicitly define `fixupPhase` ourselves and run whatever stuff _is_ relevant.

If so, that'd look something like:
```nix
fixupPhase = ''
  runHook preFixup

  # Do something...

  runHook postFixup
'';
```

---

Maybe we also want to open an issue? _Something_ in the fixup phase is buggy or inefficient; it shouldn't take 20m for any reasonable package!

Sewer56 replied:
That is true!
Ahahaha, funnily enough, I'm still not done setting up this laptop.

I'm not sure if 122993 files is a reasonable item count in a package per-se, but it is true, we probably could do better here.

MattSturgeon commented on pkgs/by-name/ca/catppuccin-papirus-folders/package.nix:
```diff
@@ -31,27 +33,43 @@ let
+  # Note(sewer56):
+  # Fetch the papirus-folders script from upstream
+  # Per instructions in the papirus-folders project.
+  # We fetch specific latest tested commit to ensure reproducibility.
```
This last sentence kinda goes without saying in nixpkgs; all external inputs are always pinned. No need for explanation.


```suggestion
```

MattSturgeon commented on pkgs/by-name/ca/catppuccin-papirus-folders/package.nix:
As per the [general commit conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) and [pkgs commit conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions), can you reword the commit to something like:

```
catppuccin-papirus-folders: 2023-08-02 -> 2024-08-06

From https://github.com/catppuccin/papirus-folders/commit/fcf96737fffc343a1bf129732c37d19f2d77fa5c
To https://github.com/catppuccin/papirus-folders/commit/f83671d17ea67e335b34f8028a7e6d78bca735d7

The most important change is the removal of the `papirus-folders` script from upstream.

Upstream now says “fetch it from papirus-folders.”
```

If there is a changelog, ideally we'd also link to it.

Sewer56 replied:
Does this imply squashing any amendments?
As the docs say `Create one commit for each logical unit.`

MattSturgeon replied:
If those amendments are fixes for one of the other commits in the PR, then yes you'd squash them.

If it is a logically distinct change, then it's better not to squash. e.g. one commit has a fix for a pre-existing issue in the package and another commit updates to a new version.

MattSturgeon commented on pkgs/by-name/ca/catppuccin-papirus-folders/package.nix:
```diff
@@ -31,27 +33,43 @@ let
+    ];
 
     postPatch = ''
+      # Copy the upstream papirus-folders script (pinned), per papirus-folders documentation
```
Personally, I would keep all explanation with the `fetchurl` bit; copying into the build directory is fairly self-explanatory.

MattSturgeon reviewed (commented):
Congratulations on your first nixpkgs PR! Looks good overall.

I'll try to find time to investigate the issue you pointed out with the fixup phase.

Sewer56 committed:
catppuccin-papirus-folders: 0-unstable-2023-08-02 -> 0-unstable-2024-08-06

From https://github.com/catppuccin/papirus-folders/commit/fcf96737fffc343a1bf129732c37d19f2d77fa5c
To https://github.com/catppuccin/papirus-folders/commit/f83671d17ea67e335b34f8028a7e6d78bca735d7

The most important change is the removal of the `papirus-folders` script from upstream.

Upstream now says “fetch it from papirus-folders.”

Sewer56 commented:
Cleaned up with the feedback, where it was possible- and of course, tested on my local machine.
All's good here.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 426978`

Logs: https://github.com/MattSturgeon/nixpkgs-review-gha/actions/runs/16662119825


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>catppuccin-papirus-folders</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>catppuccin-papirus-folders</li>
  </ul>
</details>

MattSturgeon reviewed (approved):
> commit reworded

We usually use the same format for PR titles too

---

Diff LGTM, tested build with nixpkgs-review, I haven't tested binaries as there aren't any 😁
